File: src/sparsebitvec.js
import { assert, assertNonNegative, assertSafeInteger } from "./assert.js";
import { BitBuf } from './bitbuf.js';
import * as bits from './bits.js';
import * as defaults from './defaults';
import { DenseBitVec, DenseBitVecBuilder } from './densebitvec.js';
import { IntBuf } from './intbuf.js';
import { ascending } from './sort.js';

/**
 * @implements {BitVecBuilder}
 */
export class SparseBitVecBuilder {

  /**
   * @param {number} universeSize
   */
  constructor(universeSize) {
    this.universeSize = universeSize;
    /** @type {number[]} */
    this.ones = [];
  }

  /**
   * @param {number} index
   */
  one(index, count = 1) {
    assert(index < this.universeSize, () => `index (${index}) cannot exceed universeSize (${this.universeSize})`);
    for (let i = 0; i < count; i++) {
      this.ones.push(index);
    }
  }
  
  build(options = {}) {
    this.ones.sort(ascending);
    return new SparseBitVec(this.ones, this.universeSize);
  }
}

/**
 * Sparse bitvector using Elias-Fano encoding. Supports multiplicity.
 * @implements {BitVec}
 */
export class SparseBitVec {
  /**
   * @param {number[] | Uint32Array | Float64Array} ones
   * @param {number} universeSize
   */
  constructor(ones, universeSize) {
    // disallow humungous universes because JS only supports efficient bit ops for 32-bit integers
    assert(universeSize < 2 ** 32, () => `universeSize (${universeSize}) cannot exceed 2^32 - 1`);

    // The paper "On Elias-Fano for Rank Queries in FM-Indexes" recommends a formula to compute
    // the number of low bits that is mostly equivalent to the version used below, except that
    // sometimes theirs suggests slightly worse choices, e.g. when numOnes === 25 and universeSize === 51.
    // https://observablehq.com/@yurivish/ef-split-points
    // This approach chooses the split point by noting that the trade-off effectively is between having numOnes
    // low bits, or the next power of two of the universe size separators in the high bits. Hopefully this will
    // be explained clearly in the accompanying design & background documentation.
    const numOnes = ones.length;
    const lowBitWidth = numOnes === 0 ? 0 : Math.floor(Math.log2(Math.max(1, universeSize / numOnes))); 

    // unary coding; 1 denotes values and 0 denotes separators, since that way
    // encoding becomes more efficient.
    // By default, values are never more than 50% of the bits due to the way the split point is chosen.
    // Note that this expression automatically adapts to non-power-of-two universe sizes.
    const highLength = numOnes + (universeSize >>> lowBitWidth);
    const high = new BitBuf(highLength);
    const low = new IntBuf(numOnes, lowBitWidth);
    const lowMask = bits.oneMask(lowBitWidth);

    let numUniqueOnes = 0;
    let hasMultiplicity = false;
    let prev = -1;
    for (let i = 0; i < ones.length; i++) {
      const cur = ones[i];
      hasMultiplicity ||= prev === cur;
      numUniqueOnes += Number(prev !== cur);
      assertNonNegative(cur);
      assertSafeInteger(cur);
      assert(cur < universeSize, () => `expected 1 - bit(${cur}) to not exceed the universeSize(${universeSize})`);
      assert(prev <= cur, 'expected monotonically nondecreasing sequence');
      prev = cur;

      // Encode element
      const quotient = cur >>> lowBitWidth;
      high.setOne(i + quotient);
      const remainder = cur & lowMask;
      low.push(remainder);
    }

    // todo: allow tuning of the block parameters
    // todo: explore passing a highBuilder here, so that you can eg. specify the options
    // to the bitvec type. We would have to change the above loop to use the builder, and
    // then say this.high = builder.build(buildOptions) with the options we were passed.
    /** @readonly */
    this.high = new DenseBitVec(high, 10, 10);

    /** @readonly */
    this.low = low;

    /** @readonly */
    this.numOnes = numOnes;

    /** @readonly */
    this.lowBitWidth = lowBitWidth;

    /** @readonly */
    this.lowMask = lowMask;

    /** @readonly */
    this.universeSize = universeSize;

    /** @readonly */
    this.numZeros = this.universeSize - numUniqueOnes;

    /** @readonly */
    this.hasMultiplicity = hasMultiplicity;

    /** @readonly */
    this.numUniqueOnes = numUniqueOnes;
    
    /** @readonly */
    this.numUniqueZeros = this.numZeros;
  }

  /**
   * @param {number} value
   */
  quotient(value) { 
    return value >>> this.lowBitWidth;
  }

  /**
   * @param {number} value
   */
  remainder(value) { 
    return value & this.lowMask;
  }

  /**
   * @param {number} index
   */
  rank1(index) {
    if (index < 0) {
      return 0;
    } else if (index >= this.universeSize) {
      return this.numOnes;
    }

    // Values are grouped by their upper bits, with the lower bits stored separately.
    // Compute the upper and lower bounds of the search range within the low bits,
    // which will tell us how many values within the upper bit group are below `index`.
    const quotient = this.quotient(index);
    let lowerBound = 0;
    let upperBound = 0;
    if (quotient === 0) {
      // We're searching within the first group, so the lower bound is zero.
      // Look for the divider that separates the first group from the subsequent groups.
      // If there isn't one, then we need to search the entire vector since all values
      // are in the first group.
      upperBound = this.high.trySelect0(0) ?? this.numOnes;
    } else {
      // We're searching within a higher group, so compute both the lower and the
      // upper bound from the high bit vector.
      {
        // We're searching for the i-th separator.
        // When we find it, we subtract the number of separators preceding it
        // in order to get the index of the element in the low bits.
        const i = quotient - 1;
        const n = this.high.trySelect0(i);
        lowerBound = n === null ? 0 : n - i;
      }

      {
        // Same thing, but we're searching for the next separator after that.
        const i = quotient;
        const n = this.high.trySelect0(i);
        upperBound = n === null ? this.numOnes : n - i;
      }
    }

    // Count the number of elements in this bucket that are strictly below i 
    // using just the low bits.
    const remainder = this.remainder(index);
    const bucketCount = bits.partitionPoint(upperBound - lowerBound, n => {
      const index = lowerBound + n;
      const value = this.low.get(index);
      return value < remainder;
    });

    return lowerBound + bucketCount;
  }

  /**
   * @param {number} index
   */
  rank0(index) {
    return defaults.rank0(this, index);
  }

  /**
   * @param {number} n
   */
  trySelect1(n) {
    // How many zeros are there before the nth one bit?
    const pos = this.high.trySelect1(n);
    if (pos === null) {
      return null;
    }
    const quotient = this.high.rank0(pos);
    const remainder = this.low.get(n);
    return (bits.u32(quotient << this.lowBitWidth) + remainder);
  }

  /**
   * @param {number} n
   */
  trySelect0(n) {
    const result = defaults.trySelect0(this, n);
    return result;
  }

  /**
   * @param {number} n
   */
  select1(n) {
    return defaults.select1(this, n);;
  }

  /**
   * @param {number} n
   */
  select0(n) {
    return defaults.select0(this, n);;
  };

  /**
   * @param {number} index
   */
  get(index) {
    return defaults.get(this, index);
  }
};

File: rust-playground/made-of-bits/src/sparsebitvec.rs
use crate::bitbuf::BitBuf;
use crate::bits::one_mask;
use crate::bits::partition_point;
use crate::bitvec::BitVec;
use crate::bitvec::BitVecBuilder;
use crate::densebitvec::DenseBitVec;
use crate::intbuf::IntBuf;

pub struct SparseBitVecBuilder {
    universe_size: u32,
    ones: Vec<u32>,
}

impl BitVecBuilder for SparseBitVecBuilder {
    type Target = SparseBitVec;

    fn new(universe_size: u32) -> Self {
        Self {
            universe_size,
            ones: Vec::new(),
        }
    }

    fn one_count(&mut self, bit_index: u32, count: u32) {
        if count == 0 {
            return;
        }

        assert!(bit_index < self.universe_size);
        for _ in 0..count {
            self.ones.push(bit_index);
        }
    }

    fn build(mut self) -> SparseBitVec {
        self.ones.sort();
        SparseBitVec::new(self.ones.into(), self.universe_size)
    }
}

#[derive(Clone)]
pub struct SparseBitVec {
    high: DenseBitVec,
    low: IntBuf,
    num_ones: u32,
    low_bit_width: u32,
    low_mask: u32,
    universe_size: u32,
    num_zeros: u32,
    num_unique_ones: u32,
}

impl SparseBitVec {
    pub fn new(ones: Box<[u32]>, universe_size: u32) -> Self {
        let num_ones = ones.len() as u32;

        // The paper "On Elias-Fano for Rank Queries in FM-Indexes" recommends a formula to compute
        // the number of low bits that is mostly equivalent to the version used below, except that
        // sometimes theirs suggests slightly worse choices, e.g. when numOnes === 25 and universeSize === 51.
        // https://observablehq.com/@yurivish/ef-split-points
        // This approach chooses the split point by noting that the trade-off effectively is between having numOnes
        // low bits, or the next power of two of the universe size separators in the high bits. Hopefully this will
        // be explained clearly in the accompanying design & background documentation.
        let low_bit_width = if num_ones == 0 {
            0
        } else {
            (universe_size / num_ones).max(1).ilog2()
        };

        // unary coding; 1 denotes values and 0 denotes separators, since that way
        // encoding becomes more efficient and we have a chance of saving space due to runs of
        // zeros at either end, if the values are clustered away from the domain edges.
        // By default, values are never more than 50% of the bits due to the way the split point is chosen.
        // Note that this expression automatically adapts to non-power-of-two universe sizes.
        let high_len = num_ones + (universe_size >> low_bit_width);
        let mut high = BitBuf::new(high_len);
        let mut low = IntBuf::new(num_ones, low_bit_width);
        let low_mask = one_mask(low_bit_width);

        let mut num_unique_ones = 0;
        let mut has_multiplicity = false;
        let mut prev = None;
        for (i, &cur) in ones.iter().enumerate() {
            let same = prev == Some(cur);
            num_unique_ones += if same { 0 } else { 1 };
            if let Some(prev) = prev {
                debug_assert!(prev <= cur, "ones must be sorted")
            }
            assert!(
                cur < universe_size,
                "expected 1-bit index ({}) to not exceed the universeSize ({})",
                cur,
                universe_size
            );
            debug_assert!(prev.is_none() || prev.unwrap() <= cur); // expected monotonically nondecreasing sequence
            prev = Some(cur);

            // Encode element
            let quotient = cur >> low_bit_width;
            high.set_one(i as u32 + quotient);
            let remainder = cur & low_mask;
            low.push(remainder);
        }

        let num_zeros = universe_size - num_unique_ones;
        Self {
            high: DenseBitVec::new(high, 10, 10),
            low,
            low_bit_width,
            low_mask,
            num_ones,
            num_zeros,
            num_unique_ones,
            universe_size,
        }
    }

    // todo: document this and remainder
    fn quotient(&self, x: u32) -> u32 {
        return x >> self.low_bit_width;
    }

    fn remainder(&self, x: u32) -> u32 {
        return x & self.low_mask;
    }
}

impl BitVec for SparseBitVec {
    fn rank1(&self, bit_index: u32) -> u32 {
        if bit_index >= self.universe_size() {
            return self.num_ones;
        }

        let quotient = self.quotient(bit_index);
        let lower_bound;
        let upper_bound;
        if quotient == 0 {
            // We're searching within the first group, so the lower bound is zero.
            // Look for the divider that separates the first group from the subsequent groups.
            // If there isn't one, then we need to search the entire vector since all values
            // are in the first group.
            lower_bound = 0;
            upper_bound = self.high.select0(0).unwrap_or(self.num_ones());
        } else {
            // We're searching within a higher group, so compute both the lower and the
            // upper bound from the high bit vector.

            // We're searching for the i-th separator.
            // When we find it, we subtract the number of separators preceding it
            // in order to get the index of the element in the low bits.
            let i = quotient - 1;
            let n = self.high.select0(i).map(|x| x - i);
            lower_bound = n.unwrap_or(0);

            // Same thing, but we're searching for the next separator after that.
            let i = quotient;
            let n = self.high.select0(i).map(|x| x - i);
            upper_bound = n.unwrap_or(self.num_ones());
        }

        // Count the number of elements in this bucket that are strictly below i
        // using just the low bits.

        let remainder = self.remainder(bit_index);
        let bucket_count = partition_point((upper_bound - lower_bound) as usize, |n| {
            let index = lower_bound + n as u32;
            let value = self.low.get(index);
            value < remainder
        }) as u32;

        lower_bound + bucket_count
    }

    fn select1(&self, n: u32) -> Option<u32> {
        // How many zeros are there before the nth one bit?
        let pos = self.high.select1(n)?;
        let quotient = self.high.rank0(pos);
        let remainder = self.low.get(n);
        Some((quotient << self.low_bit_width) + remainder)
    }

    fn num_ones(&self) -> u32 {
        self.num_ones
    }

    fn universe_size(&self) -> u32 {
        self.universe_size
    }

    fn num_unique_zeros(&self) -> u32 {
        self.universe_size() - self.num_unique_ones()
    }

    fn num_unique_ones(&self) -> u32 {
        self.num_unique_ones
    }

    fn get(&self, bit_index: u32) -> u32 {
        assert!(bit_index < self.universe_size());
        self.rank1(bit_index + 1) - self.rank1(bit_index)
    }

    fn has_rank0(&self) -> bool {
        !self.has_multiplicity()
    }

    fn has_select0(&self) -> bool {
        !self.has_multiplicity()
    }
}

#[cfg(test)]
mod tests {
    use crate::test_bitvec::test_bit_vec_builder;
    use crate::test_bitvec::test_bit_vec_builder_arbtest;

    use super::*;

    #[test]
    fn test() {
        test_bit_vec_builder::<SparseBitVecBuilder>();
        test_bit_vec_builder_arbtest::<SparseBitVecBuilder>(None, None, false);
        // RUST_BACKTRACE=full cargo test -- --nocapture
        // test_bit_vec_builder_arbtest::<SparseBitVecBuilder>(Some(0xac70e11d00000005), None, false);
    }
}
